### DISCUSSION TITLE
DataMember name not respected in dto generation

### DISCUSSION THREAD

**Discussion by tkat** [ServiceStack/Discuss#33] [Q&A] [answered]
ServiceStack v6 user here.

Started using TS lately, and wanted to fiddle around with the TS client. The API I'm working against, heavily utilizes "name" on attribute "DataMember". 

For instance, a member of an API response looks like this:

```csharp
[DataMember(Name = "darkMode", Order = 1)]
public string PrefersDarkMode{ get; set; }
```

JSON response is something like this:
```json
"responseObject": {
    "darkMode": "",
}
```

While it's served correctly via json resonse (property is 'darkMode'), when I try to generate types (with x tool, or by directly accessing `/types/typescript` on the api) I get a result like this:

```typescript
// @DataMember(Name="darkMode", Order=1)
public PrefersDarkMode: string;
```
The TS property name does not match the response property name (even though the serialized name is there as a comment above the type member). I checked all the code generation properties in order to find something that alters this behavior and can't seem to find something related. 

I might be missing something obvious, but all my searches regardind "servicestack" and "datamember" in forums were unsuccessful.

Thanks in advance

**Comment by mythz** [comment]
Can you provide the complete C# and generated TypeScript Request DTO please.

**Comment by tkat** [comment]
Here is an actual request.

```csharp
[Route("/sites/{Id}", "GET")]
[DataContract]
public class SiteRequest : IGet, IReturn<SiteResponse>
{
    [DataMember(Order = 1)] public long Id { get; set; }
}
```

And the generated types.

```typescript
// @Route("/sites/{Id}", "GET")
// @DataContract
export class SiteRequest implements IReturn<SiteResponse>, IGet {
  // @DataMember(Order=1)
  public id: number

  public constructor(init?: Partial<SiteRequest>) {
    ;(Object as any).assign(this, init)
  }
  public getTypeName() {
    return 'SiteRequest'
  }
  public getMethod() {
    return 'GET'
  }
  public createResponse() {
    return new SiteResponse()
  }
}
```

**Reply by mythz** [reply]
What's the issue with this?

**Reply by tkat** [reply]
There is no issue with the request, that's why I just mentioned the response and didn't include it in the first place.

**Reply by tkat** [reply]
It's just the responses that don't match DataMember name with the actual JSON property served. 

Sorry if this wasn't clear.

**Reply by tkat** [reply]
So, to be more clear, I figured that when a response has this C# property

```csharp
[DataMember(Name = "darkMode", Order = 1)]
public string PrefersDarkMode{ get; set; }
```

...and serves this property in json
```json
"darkMode": "",
```

...the type should be
```typescript
public darkMode: string;
```

... and not what's actually served

```typescript
// @DataMember(Name="darkMode", Order=1)
public prefersDarkMode: string;
```

... which is a mismatch for the actual reply, and therefore throws a TS error.

Hope this is clearer.

**Reply by mythz** [reply]
Please provide a complete C# + TypeScript DTO that contains the actual issue.

**Reply by tkat** [reply]
Here is an actual response class (C#)

```csharp
[DataContract]
public class SiteResponse
{
    [DataMember(Order = 1, Name = "id")]
    public long SiteId { get; set; }

    [DataMember(Order = 2, Name = "notificationApi")]
    public string ApiNotificationRealtimeUrl { get; set; }

    [DataMember(Order = 3, Name = "statisticsUrl")]
    public string ApiStatisticsUrl { get; set; }
}
```

... and the typescript generated types

```typescript
// @DataContract
export class SiteResponse
{
  // @DataMember(Name="id", Order=1)
  public siteId : number;

  // @DataMember(Name="notificationApi", Order=2)
  public apiNotificationRealtimeUrl: string;

  // @DataMember(Name="statisticsUrl", Order=3)
  public apiStatisticsUrl: string;
}
```

So, to summarize the problem the deserialized JSON expects three properties : id, notificationApi and statisticsUrl which are not present and thus TS throws. They are not present because instead, the C# names are preserved which is not helpful for the consuming client.

Hope this clarifies the discrepancy.

**Comment by mythz** [comment, marked as answer]
great thanks, this should now be supported from v6.3.1+ that's now [available on MyGet](https://docs.servicestack.net/myget).
